src/report/chunked.rs

```diff
@@ -0,0 +1,992 @@
+    }
+}
+
+impl From<&[u32; MAX_CHUNK_U32_WORDS]> for &WireChunk {
+    fn from(exact_array_ref: &[u32; MAX_CHUNK_U32_WORDS]) -> Self {
+        // We consider this relatively safe because WireChunk is a largely
+        // uninterpreted pile of bytes fields all with alignments == 1,
+        // and we know from static assertions that the sizes line up.
+        assert_eq_size!([u32; MAX_CHUNK_U32_WORDS], WireChunk);
+        assert_eq_align!(u8, WireChunk);
+        unsafe { &*(exact_array_ref.as_ptr() as *const WireChunk) }
+    }
+}
+
+/// The slice input was an incorrect length.
+#[derive(Debug, PartialEq, Eq)]
+pub struct IncorrectLengthSlice;
+
+/// The things that can go wrong when writing a chunked report.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum ChunkedReportError {
+    /// One of the general-case errors that can occur
+    /// in producing a report.
+    ReportError(crate::ReportError),
+    /// No chunked report transaction has been started.
+    NoChunkedReportInProgress,
+}
+
+/// Correlation value threaded through the steps of a chunked
+/// report in order to prevent data corruption.
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq)]
+pub struct ChunkedReportToken {
+    /// The identifier for this report / group-of-chunks
+    group_id: u16,
+}
+
+/// Write reports in 1+ chunks.
+///
+/// Call `start_chunked_report` once at the start of a report.
+/// Call `write_next_report_chunk` as many times as needed to
+/// produce all the report pieces, transmitting each piece as you go.
+/// Call `finish_chunked_report` when all the chunks have been made.
+pub trait ChunkedReporter {
+    /// Prepare for chunked reporting, beginning a chunked reporting transaction.
+    /// Returns a token that should be re-used within this chunked reporting transaction.
+    ///
+    /// Lock out instance internal mutation to a degree necessary to prevent data corruption.
+    fn start_chunked_report(&mut self) -> Result<ChunkedReportToken, ChunkedReportError>;
+
+    /// Write up to MAX_CHUNK_BYTES bytes of the chunked format into the provided buffer.
+    ///
+    /// Return number of bytes written on success.
+    ///
+    /// A successful return value of 0 bytes indicates that no payload bytes could be written
+    /// because all chunks for this report transaction have been handled.
+    fn write_next_report_chunk(
+        &mut self,
+        token: &ChunkedReportToken,
+        destination: &mut [u8],
+    ) -> Result<usize, ChunkedReportError>;
+
+    /// Idempotent cessation and cleanup of a (possibly error'd out)
+    /// chunked reporting transaction.
+    /// Undo any internal mutation locks related to reporting.
+    fn finish_chunked_report(
+        &mut self,
+        token: ChunkedReportToken,
+    ) -> Result<(), ChunkedReportError>;
+}
+
+#[derive(Debug)]
+pub(crate) struct ChunkedReportState {
+    /// Use the None-value to indicate "no chunked report in progress"
```

Comment:
There's no Option here, did you mean 'false'?

Reply:
Yep, that's a remnant of a prior version, will correct.